core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -112,6 +137,11 @@ public ExpireSnapshots cleanExpiredFiles(boolean clean) {
   @Override
   public ExpireSnapshots expireSnapshotId(long expireSnapshotId) {
     LOG.info("Expiring snapshot with id: {}", expireSnapshotId);
+    List<Map.Entry<String, SnapshotRef>> refsForSnapshot = base.refs().entrySet().stream()
+        .filter(refEntry -> refEntry.getValue().snapshotId() == expireSnapshotId)
+        .collect(Collectors.toList());
+    Preconditions.checkArgument(refsForSnapshot.isEmpty(), "Cannot expire %d. There are %d references to it",
+        expireSnapshotId, refsForSnapshot.size());
```

Comment:
Wrong place to put this check, we should do it in the internalApply after computing which refs will be retained. then we can check if the snapshots of the refs to be retained contains any of the individual snapshots in idsToRemove.